src/proxyp/Message.cc

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "proxyp/Elements.h"
+#include "proxyp/Message.h"
+#include "sbuf/Stream.h"
+#include "sbuf/StringConvert.h"
+#include "SquidConfig.h"
+#include "StrList.h"
+
+ProxyProtocol::Message::Message(const char *ver, const uint8_t cmd):
+    version_(ver),
+    command_(Two::Command(cmd)),
+    ignoreAddresses_(false)
+{}
+
+SBuf
+ProxyProtocol::Message::toMime() const
+{
+    SBufStream result;
+    for (const auto &p: PseudoHeaderFields)
+        result << p.first << ": " << getValues(p.second) << "\r\n";
```

Comment:
I do not think we should add lines with empty pseudo header values. Please refactor to skip lines with an empty getValues() result. At that point, it may make sense to manually enumerate pseudo headers (as an optimization), so that we can skip all addresses/ports at once.